### PR TITLE
interfaces: reconnect, fix for autoconnect/reconnect conflict check

### PR DESCRIPTION
**Pull request by stolowski**
Use the helper introduced in #5196 to fix reconnect and autoconnect logic not to get into infinite retires if both core and a snap connected to core is refreshed at the same time. Re-worked doAutoconnect and doReconnect logic to check for conflicts first, before creating tasks to avoid the problem of creating multiple hook tasks with no associated change if an error occurs when creating them.
Also changed the internals of connect to delay the creation of task and determine initial hook attribiutes first (and error out early before creating hook tasks). Note, the ConnectOnInstall method doesn't check for conflicts internally now, instead the conflicts needs to be checked for upfront.

Marking blocked as the helper needs to land first.
Also, not addressing the problem of suboptimal autoconnect handling where we run hooks twice for both ends  - will to in a followup.

### CODE REVIEW

File: overlord/ifacestate/handlers.go [hunk 1]
```diff
@@ -518,6 +518,51 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
...
+
+	// Check for conflicts first and only create all tasks if we're clear of conflicts
+	chg := task.Change()
+	for _, conn := range connections {
+		if err := checkConnectConflicts(st, chg, conn.PlugRef.Snap, conn.SlotRef.Snap, task, waitingOnUs); err != nil {
+			return err
+		}
+	}
+
+	task.SetStatus(state.DoneStatus)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/ifacestate/handlers.go]: Indeed, it makes sense to do it later, thanks.

File: overlord/ifacestate/handlers.go [hunk 2]
```diff
@@ -691,17 +738,23 @@ func (m *InterfaceManager) doAutoConnect(task *state.Task, _ *tomb.Tomb) error {
...
-				continue
+			if err := checkConnectConflicts(st, chg, plug.Snap.Name(), slot.Snap.Name(), task, waitingOnUs); err != nil {
+				return err
 			}
-			autots.AddAll(ts)
+			newconns = append(newconns, connRef)
 		}
 	}
 
 	task.SetStatus(state.DoneStatus)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/ifacestate/handlers.go]: Done.

File: overlord/ifacestate/ifacestate.go [hunk 3]
```diff
@@ -101,22 +108,23 @@ func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slot
 	return nil
 }
 
-// ConnectOnInstall returns a set of tasks for (re)connecting an interface as part of snap installation (e.g. to reconnect interfaces on refresh, or autoconnect)
+// ConnectOnInstall returns a set of tasks for (re)connecting an interface as part of snap installation (e.g. to reconnect interfaces on refresh, or autoconnect).
+// This method doesn't check for conflicts, they need to be checked for upfront.
```
Comment (pedronis) [inline overlord/ifacestate/ifacestate.go]: given it does not conflicts check this shouldn't be exported I think,  we should just use connect directly

File: overlord/ifacestate/handlers.go [hunk 4]
```diff
@@ -655,12 +704,10 @@ func (m *InterfaceManager) doAutoConnect(task *state.Task, _ *tomb.Tomb) error {
 			continue
 		}
 
-		ts, err := ConnectOnInstall(st, chg, task, plug.Snap.Name(), plug.Name, slot.Snap.Name(), slot.Name)
-		if err != nil {
-			task.Logf("cannot auto-connect plug %s to %s: %s", connRef.PlugRef, connRef.SlotRef, err)
-			continue
+		if err := checkConnectConflicts(st, chg, plug.Snap.Name(), slot.Snap.Name(), task, waitingOnUs); err != nil {
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: as discussed we probably want to go back to ignore conflicts and errors here, so here we should ignore conns with conflicts (but probably report other kinds of errors)

File: overlord/ifacestate/ifacestate.go [hunk 5]
```diff
@@ -160,13 +173,16 @@ func connect(st *state.State, change *state.Change, mainTask *state.Task, plugSn
...
 
 	// Expose a copy of all plug and slot attributes coming from yaml to interface hooks. The hooks will be able
 	// to modify them but all attributes will be checked against assertions after the hooks are run.
-	if err := setInitialConnectAttributes(connectInterface, plugSnap, plugName, slotSnap, slotName); err != nil {
-		return nil, err
-	}
+	emptyDynamicAttrs := map[string]interface{}{}
+	connectInterface.Set("plug-static", plugStatic)
+	connectInterface.Set("slot-static", slotStatic)
+	connectInterface.Set("plug-dynamic", emptyDynamicAttrs)
```
Comment (pedronis) [inline overlord/ifacestate/ifacestate.go]: as we discussed with this is a bit unclear how the snaps can distinguish initial connect vs reconnect and reuse/avoid leaking resources  without rolling their own handshake and bookkeeping state

**Comment by stolowski** [comment]
Closing as discussed, no reconnect for now.

**Pull request closed without merging**